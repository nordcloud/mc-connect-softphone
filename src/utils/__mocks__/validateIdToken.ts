export const validateIdToken = jest.fn().mockResolvedValue({
  active: true,
  sub: 'dummySub',
  name: 'dummyName',
  email: 'dummyEmail',
});
