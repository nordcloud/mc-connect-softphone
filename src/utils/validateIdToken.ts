import fetch from 'node-fetch';

import { OauthIntrospectResponse } from '../types';
import { getStageConsts } from './getStageConsts';

const { OAUTH_BASE_URL, OAUTH_CLIENT_ID } = getStageConsts();

export async function validateIdToken(idToken: string) {
  console.log('Validating ID token');

  const res = await fetch(`${OAUTH_BASE_URL}/introspect`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      client_id: OAUTH_CLIENT_ID,
      token: idToken,
      token_type_hint: 'id_token',
    }),
  });

  const data = (await res.json()) as OauthIntrospectResponse;

  if ('error' in data) {
    throw new Error(`${data.error}: ${data.error_description}`);
  }

  return data;
}
