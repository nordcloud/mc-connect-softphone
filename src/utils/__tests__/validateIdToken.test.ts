import fetch, { Response } from 'node-fetch';

import { validateIdToken } from '../validateIdToken';

jest.mock('node-fetch');

jest.mocked(fetch).mockResolvedValue({
  json: () => Promise.resolve({}),
} as Response);

const token = 'dummyIdToken';

describe('validateIdToken', () => {
  it('calls fetch with correct args', async () => {
    await validateIdToken(token);
    expect(jest.mocked(fetch).mock.calls).toMatchSnapshot();
  });

  it('rejects with a correct value when "fetch" response has "error" prop', () => {
    jest.mocked(fetch).mockResolvedValueOnce({
      json: () =>
        Promise.resolve({
          error: 'dummyError',
          error_description: 'dummyErrorDescription',
        }),
    } as Response);
    return expect(validateIdToken(token)).rejects.toMatchSnapshot();
  });
});
