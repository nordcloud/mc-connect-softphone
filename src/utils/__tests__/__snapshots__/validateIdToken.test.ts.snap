// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateIdToken calls fetch with correct args 1`] = `
[
  [
    "https://nordcloud.okta.com/oauth2/v1/introspect",
    {
      "body": URLSearchParams {
        Symbol(query): [
          "client_id",
          "0oaqyrebtuY1eagfS357",
          "token",
          "dummyIdToken",
          "token_type_hint",
          "id_token",
        ],
        Symbol(context): null,
      },
      "headers": {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`validateIdToken rejects with a correct value when "fetch" response has "error" prop 1`] = `[Error: dummyError: dummyErrorDescription]`;
